File: csharp/Card.cs
using System.Collections.Generic;
using System.Linq;

namespace Yuizumi.TexasHoldem
{
    public enum Rank
    {
        _2, _3, _4, _5, _6, _7, _8, _9, _T, _J, _Q, _K, _A,
    }

    public enum Suit
    {
        Spades, Hearts, Diamonds, Clubs,
    }

    public sealed class Card
    {
        private const string RankChars = "23456789TJQKA";
        private const string SuitChars = "SHDC";

        private Card(int index)
        {
            Index = index;
        }

        static Card()
        {
            Deck = Enumerable.Range(0, 52).Select(index => new Card(index))
                .ToList().AsReadOnly();
        }

        public static IReadOnlyList<Card> Deck { get; }

        public int Index { get; }

        internal int RankIndex
        {
            get { return Index >> 2; }
        }

        internal int SuitIndex
        {
            get { return Index & 3; }
        }

        public Rank Rank
        {
            get { return (Rank) RankIndex; }
        }

        public Suit Suit
        {
            get { return (Suit) SuitIndex; }
        }

        public static Card Parse(string value)
        {
            if (value.Length != 2)
                return null;
            int rank = RankChars.IndexOf(value[0]);
            int suit = SuitChars.IndexOf(value[1]);
            if (rank == -1 || suit == -1)
                return null;
            return Deck[(rank << 2) | suit];
        }

        public override string ToString()
        {
            return $"{RankChars[RankIndex]}{SuitChars[SuitIndex]}";
        }
    }
}

File: cpp/card.cc
#include "card.h"

#include <cassert>
#include <cstring>

using std::string;
using std::vector;

namespace texas_holdem {

namespace {
const char kRankChars[] = "23456789TJQKA";
const char kSuitChars[] = "SHDC";
}  // namespace

bool Card::Parse(const string& name, Card* card)
{
    if (name.length() != 2)
        return false;
    const char* rank = strchr(kRankChars, name[0]);
    const char* suit = strchr(kSuitChars, name[1]);
    if (!rank || !suit)
        return false;
    card->rank_ = static_cast<Rank>(rank - kRankChars);
    card->suit_ = static_cast<Suit>(suit - kSuitChars);
    return true;
}

string Card::ToString() const
{
    char name[2] = {kRankChars[rank()], kSuitChars[suit()]};
    return string(name, 2);
}

const vector<Card>& Card::Deck()
{
    static const vector<Card> deck(CreateDeck());
    return deck;
}

vector<Card> Card::CreateDeck()
{
    const int num_cards = kNumRanks * kNumSuits;
    vector<Card> deck;
    deck.reserve(num_cards);
    for (int rank = 0; rank < kNumRanks; rank++) {
        for (int suit = 0; suit < kNumSuits; suit++) {
            deck.push_back(Card(static_cast<Rank>(rank),
                                static_cast<Suit>(suit)));
            assert(deck.back().index() == static_cast<int>(deck.size()) - 1);
        }
    }
    return deck;
}

}  // namespace texas_holdem
